File: android/app/src/test/java/com/fitlog/app/FormulasTest.kt
package com.fitlog.app

import com.fitlog.app.domain.Formulas
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class FormulasTest {

    @Test
    fun `el volumen coincide con los vectores compartidos`() {
        val vectors = TestVectors.load("formulas.json").jsonObject.getValue("volume").jsonArray
        for (case in vectors) {
            val entry = case.jsonObject
            val weightKg = entry.getValue("weight_kg").jsonPrimitive.doubleOrNull
            val reps = entry.getValue("reps").jsonPrimitive.intOrNull
            val expected = entry.getValue("expected").jsonPrimitive.doubleOrNull

            val actual = Formulas.volumeKg(weightKg, reps)
            if (expected == null) {
                assertNull("volumen($weightKg, $reps) deberia ser null", actual)
            } else {
                assertEquals("volumen($weightKg, $reps)", expected, actual ?: Double.NaN, 1e-9)
            }
        }
    }

    @Test
    fun `el 1RM estimado coincide con los vectores compartidos`() {
        val vectors = TestVectors.load("formulas.json")
            .jsonObject.getValue("one_rep_max_epley").jsonArray
        for (case in vectors) {
            val entry = case.jsonObject
            val weightKg = entry.getValue("weight_kg").jsonPrimitive.doubleOrNull
            val reps = entry.getValue("reps").jsonPrimitive.intOrNull
            val expected = entry.getValue("expected").jsonPrimitive.doubleOrNull

            val actual = Formulas.estimatedOneRepMaxKg(weightKg, reps)
            if (expected == null) {
                assertNull("1RM($weightKg, $reps) deberia ser null", actual)
            } else {
                assertEquals("1RM($weightKg, $reps)", expected, actual ?: Double.NaN, 1e-9)
            }
        }
    }
}

File: web/src/test/formulas.test.ts
import { describe, expect, it } from 'vitest';

import { estimatedOneRepMaxKg, volumeKg } from '@/domain/formulas';

import vectors from '@shared/test-vectors/formulas.json';

interface VolumeCase {
  weight_kg: number | null;
  reps: number | null;
  expected: number | null;
}

interface OneRepMaxCase {
  weight_kg: number | null;
  reps: number | null;
  expected: number | null;
}

const volumeCases: VolumeCase[] = vectors.volume;
const oneRepMaxCases: OneRepMaxCase[] = vectors.one_rep_max_epley;

describe('volumeKg', () => {
  it.each(volumeCases)(
    'peso $weight_kg x $reps repeticiones -> $expected',
    ({ weight_kg, reps, expected }) => {
      expect(volumeKg(weight_kg, reps)).toBe(expected);
    }
  );
});

describe('estimatedOneRepMaxKg (Epley)', () => {
  it.each(oneRepMaxCases)(
    'peso $weight_kg x $reps repeticiones -> $expected',
    ({ weight_kg, reps, expected }) => {
      const result = estimatedOneRepMaxKg(weight_kg, reps);
      if (expected === null) {
        expect(result).toBeNull();
        return;
      }
      expect(result).not.toBeNull();
      expect(result as number).toBeCloseTo(expected, 9);
    }
  );
});
